tuned/plugins/hotplug.py

```diff
@@ -17,42 +17,68 @@ def cleanup(self):
 		self._hardware_events_cleanup()
 
 	def _hardware_events_init(self):
-		raise NotImplementedError()
+		pass
 
 	def _hardware_events_cleanup(self):
-		raise NotImplementedError()
+		pass
 
 	def _init_devices(self):
 		self._hardware_events_init()
 
 	def _hardware_events_callback(self, event, device):
 		if event == "add":
 			log.info("device '%s' added" % device.sys_name)
-			self._add_device(device)
+			self._add_device(device.sys_name)
 		elif event == "remove":
 			log.info("device '%s' removed" % device.sys_name)
-			self._remove_device(device)
+			self._remove_device(device.sys_name)
 
-	def _add_device(self, device):
-		device_name = device.sys_name
+	def _add_device_process(self, instance_name, instance, device_name):
+		log.info("instance %s: adding new device %s" % (instance_name, device_name))
+		self._assigned_devices.add(device_name)
+		self._call_device_script(instance, instance.script_pre, "apply", [device_name])
+		self._added_device_apply_tuning(instance, device_name)
+		self._call_device_script(instance, instance.script_post, "apply", [device_name])
+		instance.processed_devices.add(device_name)
+
+	def _add_device(self, device_name):
 		if device_name in (self._assigned_devices | self._free_devices):
 			return
 
 		for instance_name, instance in list(self._instances.items()):
 			if len(self._get_matching_devices(instance, [device_name])) == 1:
-				log.info("instance %s: adding new device %s" % (instance_name, device_name))
-				self._assigned_devices.add(device_name)
-				self._call_device_script(instance, instance.script_pre, "apply", [device_name])
-				self._added_device_apply_tuning(instance, device_name)
-				self._call_device_script(instance, instance.script_post, "apply", [device_name])
-				instance.processed_devices.add(device_name)
+				self._add_device_process(instance_name, instance, device_name)
 				break
 		else:
 			log.debug("no instance wants %s" % device_name)
 			self._free_devices.add(device_name)
 
-	def _remove_device(self, device):
-		device_name = device.sys_name
+	def _add_device_to_instance(self, device_name, instance_name):
+		"""
+		Add device to specific instance, no check is performed whether device is compatible with the instance
+		or accepted by the instance.
+		"""
+		if device_name in self._assigned_devices:
+			log.error("device '%s' is already assigned to instance" % device_name)
```

Comment:
Would it make sense to drop the severity to Warning here?  Error seems too high.